qiskit/providers/aer/backends/aer_compiler.py

```diff
@@ -40,27 +41,29 @@ def compile(self, circuits, basis_gates=None, optypes=None):
                             (default: None).
 
         Returns:
-            QuantumCircuit or list: QuantumCircuit (or list
-                of QuantumCircuit objects) without control-flow instructions
+            tuple: Typle of list QuantumCircuit without control-flow
+                   instructions and list of optypes for each circuit.
         """
-        if isinstance(circuits, QuantumCircuit):
+        if not isinstance(circuits, list):
             circuits = [circuits]
+        if optypes is None:
+            optypes = len(circuits) * [None]
         if isinstance(circuits, list):
             basis_gates = basis_gates + ['mark', 'jump']
             compiled_circuits = []
             for idx, circuit in enumerate(circuits):
-                optype = optypes[idx] if optypes else None
-                if isinstance(circuit, QuantumCircuit) and self._is_dynamic(
-                        circuit, optype):
+                if self._is_dynamic(circuit, optypes[idx]):
                     compiled_circ = transpile(
                         self._inline_circuit(circuit, None, None),
                         basis_gates=basis_gates
                     )
                     compiled_circuits.append(compiled_circ)
+                    # Recompute optype for compiled circuit
+                    optypes[idx] = circuit_optypes(compiled_circ)
                 else:
                     compiled_circuits.append(circuit)
-            return compiled_circuits
-        return circuits
+            return compiled_circuits, optypes
+        return circuits, optypes
```

Comment:
This is 100% a public API change, and I'd be super nervous about just changing the return type like this.  You can get around it simply by adding a `return_optypes` keyword argument to the function, and return `compiled_circuits` or `circuits` (the previous behaviour) only, unless it's specified as `True`, and update calling library code to use that.

(Putting this in now is how we change it safely - in the next Aer release you can make it raise a deprecation warning if the value isn't given, etc etc.)

Reply:
this function isn't and everything in this module isn't included in any init files or import paths, so its not really public

Reply:
Changed this to return just circuits in cases where optypes is None